### PR TITLE
Update request-resolve mechanism for full archive network

### PR DESCRIPTION
**Pull request by sstanculeanu**
## Reasoning behind the pull request
- request-resolve mechanism should be updated for the full archive network
  
## Proposed changes
- updated resolvers and requesters
- added integration test on peers ratings

## Testing procedure
- will be tested with feat branch

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: factory/network/networkComponentsHandler.go [hunk 1]
```diff
@@ -259,6 +259,18 @@ func (mnc *managedNetworkComponents) FullArchivePeersRatingMonitor() p2p.PeersRa
...
+// FullArchivePreferredPeersHolderHandler returns the preferred peers holder of the full archive network
+func (mnc *managedNetworkComponents) FullArchivePreferredPeersHolderHandler() factory.PreferredPeersHolderHandler {
+	mnc.mutNetworkComponents.RLock()
+	defer mnc.mutNetworkComponents.RUnlock()
+
+	if mnc.networkComponents == nil {
+		return nil
+	}
+
+	return mnc.mainNetworkHolder.preferredPeersHolder
```
Comment (ssd04) [inline factory/network/networkComponentsHandler.go:271]:
```suggestion
func (mnc *managedNetworkComponents) FullArchivePreferredPeersHolderHandler() factory.PreferredPeersHolderHandler {
	mnc.mutNetworkComponents.RLock()
	defer mnc.mutNetworkComponents.RUnlock()

	if mnc.networkComponents == nil {
		return nil
	}

	return mnc.fullArchiveNetworkHolder.preferredPeersHolder
```
?

File: factory/network/networkComponentsHandler.go [hunk 2]
```diff
@@ -259,6 +259,18 @@ func (mnc *managedNetworkComponents) FullArchivePeersRatingMonitor() p2p.PeersRa
 	return mnc.fullArchiveNetworkHolder.peersRatingMonitor
 }
 
+// FullArchivePreferredPeersHolderHandler returns the preferred peers holder of the full archive network
+func (mnc *managedNetworkComponents) FullArchivePreferredPeersHolderHandler() factory.PreferredPeersHolderHandler {
```
Comment (iulianpascalau) [inline factory/network/networkComponentsHandler.go:263]: I do not think this is required

File: integrationTests/p2p/peersRating/peersRating_test.go [hunk 3]
```diff
@@ -111,12 +113,103 @@ func TestPeersRatingAndResponsiveness(t *testing.T) {
 	assert.Equal(t, finalMaliciousExpectedRating, maliciousRating)
 }
 
-func createNodeWithPeersRatingHandler(shardID uint32, numShards uint32) *integrationTests.TestProcessorNode {
+func TestPeersRatingAndResponsivenessOnFullArchive(t *testing.T) {
```
Comment (iulianpascalau) [inline integrationTests/p2p/peersRating/peersRating_test.go:116]: :+1: 
very good test for ensuring that the feature works

File: factory/network/networkComponentsHandler.go (hunk 1 above)
Comment (sstanculeanu) [reply factory/network/networkComponentsHandler.go:271]: updated, good catch

File: cmd/node/config/prefs.toml [hunk 4]
```diff
@@ -28,6 +28,15 @@
...
    PreferredConnections = []
 
+   # PreferredFullArchiveConnections holds an array containing valid ips or peer ids from nodes to connect with (in top of other connections)
+   # This is only considered on FullArchive mode but each full archive preferred peer must be added to PreferredConnections as well
+   # Example:
+   # PreferredConnections = [
+   #    "127.0.0.10",
+   #    "16Uiu2HAm6yvbp1oZ6zjnWsn9FdRqBSaQkbhELyaThuq48ybdorrr"
+   # ]
+   PreferredFullArchiveConnections = []
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply cmd/node/config/prefs.toml:38]: removed and reused the old one

File: dataRetriever/topicSender/topicResolverSender.go [hunk 5]
```diff
@@ -31,7 +31,11 @@ func NewTopicResolverSender(arg ArgTopicResolverSender) (*topicResolverSender, e
 // Send is used to send an array buffer to a connected peer
 // It is used when replying to a request
 func (trs *topicResolverSender) Send(buff []byte, peer core.PeerID) error {
-	return trs.sendToConnectedPeer(trs.topicName, buff, peer)
+	if trs.fullArchiveMessenger.IsConnected(peer) {
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply dataRetriever/topicSender/topicResolverSender.go:34]: added TODO and will be fixed in the next PR, as discussed

**Review by ssd04 — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**